drivers: ethernet: Add Jailhouse IVSHMEM Ethernet support

gramsay0 opened the pull request:
Allows Ethernet communication between "cells" in the Jailhouse hypervisor.

This has been tested with Jailhouse running inside QEMU Cortex-A53 (Zephyr <-> Linux), and without Jailhouse by running two QMU x86_64 instances (Zephyr <-> Zephyr) communicating over an ivshmem-v2 server (from the Siemens QEMU fork).
It _should_ only require device tree modifications to run on real hardware.

gramsay0 commented on samples/drivers/ethernet/eth_ivshmem/boards/qemu_cortex_a53.overlay:
```diff
@@ -0,0 +1,46 @@
+};
+
+&pcie {
+	eth_ivhsmem0: eth_ivhsmem@0 {
```
The device tree does not currently have a nice way of dealing with PCI INTx interrupt configuration...

To get this working:

- `eth_ivhsmem0` had to be a child of `pcie`, rather than just `interrupt-parent = <&pcie>;`
- A dummy register address/size was required `reg = <0x00 0x00 0x00 0x00 0x01>;`
- All four INTx interrupts are allocated to allow runtime lookup
- Additional code in the Ethernet driver to lookup the correct interrupt

I think a better solution would be to modify the device tree code to auto-generate the interrupt-map inside the PCI controller driver. This would simplify INTx interrupts usage, reduce duplication and remove the need for these DTS kludges

carlescufi commented:
@tbursztyka it'd be great if you could review this one

tbursztyka commented on drivers/ethernet/eth_ivshmem.c:
```diff
@@ -0,0 +1,594 @@
+		shmem_phys_addr = mbar_shmem.phys_addr;
+	} else {
+		cap_pos = vendor_cap + IVSHMEM_CFG_ADDRESS / 4;
+		shmem_phys_addr = pcie_conf_read_u64(pcie.bdf, cap_pos);
```
what specs are you following here? at least it is not in original ivshmem specs.

gramsay0 replied:
Jailhouse "ivshmem-v2":

https://github.com/siemens/jailhouse/blob/630001202caec85370fb4f956e581f51109e490c/Documentation/ivshmem-v2-specification.md

tbursztyka commented on drivers/ethernet/eth_ivshmem.c:
```diff
@@ -0,0 +1,594 @@
+struct eth_ivshmem_cfg_data {
+	const char *name;
+	void (*generate_mac_addr)(uint8_t mac_addr[6]);
+	struct IntxInfo {
```
CamelCase is not part of zephyr code style. Change that.

tbursztyka reviewed (changes requested):
Why haven't you followed the generic ivshmem API?
(and, if necessary, upgraded the driver so it works on non MSI-X?)

Ivshmem, by definition, is just a technic to share memory and provide interrupts.

So you get the generic logic in one place (ivshmem driver and API) and specialized communication pipes on top.


This ethernet driver should be built on top of ivshmem API. And if there is any issue with the existing API: let's fix it.

tbursztyka commented on drivers/ethernet/eth_ivshmem.c:
```diff
@@ -0,0 +1,594 @@
+		return -ENODEV;
+	}
+
+	LOG_INF("PCIe: ID 0x%08X, BDF 0x%X", pcie.id, pcie.bdf);
```
all the above should not be there. See DEVICE_PCIE_INST_DECLARE() instead

gramsay0 replied:
Currently that functionality only allows devices to be found using a vendor/device-id tuple.

There can be multiple ivshmem PCI devices present in the controller, only those with `IVSHMEM_PROTO_NET` in the programming interface register (ProgIf) are suitable for this driver.

I used a PCI scan to do this, the alternative would be to add additional device lookup parameters to the PCI controller

gramsay0 replied:
@tbursztyka I just added PCI device filtering on the class-rev register to tidy this up

gramsay0 commented:
> Why haven't you followed the generic ivshmem API? (and, if necessary, upgraded the driver so it works on non MSI-X?)
> 
> Ivshmem, by definition, is just a technic to share memory and provide interrupts.
> 
> So you get the generic logic in one place (ivshmem driver and API) and specialized communication pipes on top.
> 
> This ethernet driver should be built on top of ivshmem API. And if there is any issue with the existing API: let's fix it.

@tbursztyka Jailhouse uses "ivshmem-v2", I'm not sure how compatible this is with ivshmem.
I'll have a look and get back to you on this

https://github.com/siemens/jailhouse/blob/630001202caec85370fb4f956e581f51109e490c/Documentation/ivshmem-v2-specification.md
https://github.com/siemens/jailhouse/blob/630001202caec85370fb4f956e581f51109e490c/Documentation/inter-cell-communication.md

gramsay0 commented:
@tbursztyka I've had a look and the specs for ivshmem/ivshmem-v2 seem quite different. Here are some of the changes from [an oldish article](https://lists.gnu.org/archive/html/qemu-devel/2017-01/msg02860.html):

- Side-band registers to discover and configure share memory regions
- INTx interrupts support is back
- Multiple shared memory regions, including unidirectional ones
- Life-cycle management via local and remote state 

It looks like adding ivshmem-v2 to QEMU is still pending:
https://patchwork.kernel.org/project/qemu-devel/cover/cover.1573477032.git.jan.kiszka@siemens.com/
https://patchwork.kernel.org/project/qemu-devel/cover/cover.1578407802.git.jan.kiszka@siemens.com/

I could have a look into creating a separate ivshmem-v2 API if you think that is beneficial?

tbursztyka commented:
> I could have a look into creating a separate ivshmem-v2 API if you think that is beneficial?

Ok, did not know about that new version! Thanks for the info

Sure this would be a nice addition. Actually, make a diff of the features between v1 and v2. It's probably not necessary to have a separate API, but instead an improved ivshmem API that could deal with both versions.

gramsay0 committed:
drivers: pcie: Enable filtering PCIe devices by class-rev

This allows finding the correct PCIe device when multiple devices
have the same vendor-id/device-id but differ in the class-rev register

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 commented:
This PR has grown rather large...
I can split it up if that will help reviewers

@tbursztyka I've added ivshmem-v2 support to the existing ivshmem API

gramsay0 commented:
Hey just giving this a bump.
Any other thoughts/reviews?

gramsay0 commented:
Another bump...

tbursztyka commented on drivers/virtualization/Kconfig:
```diff
@@ -28,10 +28,9 @@ source "subsys/logging/Kconfig.template.log_config"
 	help
 	  This will enable support of ivshmem-doorbell, i.e. the interrupt
-	  based ivshmem.
+	  based ivshmem. For ivshmem-v2 INTx interrupts are also supported.
```
Have a new config option called IVSHMEM_V2

tbursztyka commented on include/zephyr/drivers/virtualization/ivshmem.h:
```diff
@@ -36,19 +39,47 @@ typedef int (*ivshmem_register_handler_f)(const struct device *dev,
 	ivshmem_get_vectors_f get_vectors;
 	ivshmem_int_peer_f int_peer;
 	ivshmem_register_handler_f register_handler;
+	/* ivshmem-v2 */
```
The API should be available only if CONFIG_IVSHMEM_V2 is enabled.

tbursztyka commented on drivers/virtualization/virt_ivshmem.h:
```diff
@@ -25,13 +46,27 @@ struct ivshmem_param {
 struct ivshmem {
 	DEVICE_MMIO_RAM;
 	struct pcie_dev *pcie;
+	bool ivshmem_v2;
```
#ifdef CONFIG_IVSHMEM_V2

gramsay0 replied:
Thanks for giving this a futher review.

These were my thoughts when implementation this:

* for completeness, this runtime flag would allow using both ivshmem and ivshmem-v2 concurrently 
* the class of processor that use this functionality are probably less concerned with the memory usage of a few unused functions
* keeping the runtime flag and adding the ifdef makes the implementation a little messy

Do you think this should be runtime configurable, ifdef or both?

tbursztyka commented on drivers/virtualization/virt_ivshmem.h:
```diff
@@ -25,13 +46,27 @@ struct ivshmem_param {
 	struct ivshmem_param params[CONFIG_IVSHMEM_MSI_X_VECTORS];
 	uint16_t n_vectors;
 #endif
+	uint32_t max_peers;
+	size_t rw_section_size;
+	size_t rw_section_offset;
+	size_t output_section_size;
+	size_t output_section_offset;
```
same

tbursztyka commented on drivers/virtualization/virt_ivshmem.h:
```diff
@@ -41,6 +76,14 @@ struct ivshmem_reg {
 	uint32_t doorbell;
 };
 
+struct ivshmem_v2_reg {
```
same etc...

tbursztyka reviewed (changes requested):
Sorry for the delay,

split the commit about ivshmem v2 API in 2: 
- one with kconfig and header file modification
- another one with the actual implementation 

Applying my comments as well (imo ivshmem-v2 should be optional to ivshmem support)

About your ethernet driver, as it is a new driver: directly make it using the ivshmem-v2 API. Less commits and intermediary states.

Besides that, all looks fine to me.

gramsay0 committed:
drivers: virtualization: Add interface for ivshmem-v2

ivshmem-v2 is primarily used for IPC in the Jailhouse hypervisor

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 committed:
drivers: virtualization: Add implementation for ivshmem-v2

ivshmem-v2 is primarily used for IPC in the Jailhouse hypervisor

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 committed:
drivers: ethernet: Add Jailhouse IVSHMEM Ethernet support

Allows Ethernet communication between "cells"
in the Jailhouse hypervisor.

The vring queue deviates from a standard virtqueue
so is implemented separately.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 committed:
tests: drivers: ethernet: Add eth_ivshmem_queue unit tests

These tests were useful during development,
and may be useful to find/fix bugs.

OpenAMP does not currently support native_posix
boards so qemu_cortex_a53 is used.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 committed:
samples: drivers: ethernet: Add Jailhouse IVSHMEM Ethernet sample

Add an IVSHMEM Ethernet sample to communicate between Zephyr
and Linux root cell within the Jailhouse hypervisor.

The supported configuration runs the Jailhouse hypervisor
within QEMU Cortex-A53.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 committed:
arch: arm64: cache: Add ARG_UNUSED to unused arguments

The unused arguments were causing warnings

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

cfriedt commented:
@jgl-meta - this is quite cool stuff.
